[TF-31901] Update Explorer API documentation to add Resource type (#2599)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/api-docs/explorer.mdx

| Parameter      | Description |
| -------------- | ----------------------------------------------------------------------------- |

> | `type`         | **Required.** Must be one of the following available views: `workspaces`, `tf_versions`, `providers`, `modules` or `resources`. See the [View Types](#view-types) section for more information about each. |

Reviewer: ```suggestion
| `type`         | **Required.** Must be one of the following available views: `workspaces`, `tf_versions`, `providers`, `modules`, or `resources`. See the [View Types](#view-types) section for more information about each. |
```
We always use the serial comma

| `tf_versions` | Information about the Terraform versions in use across this organization. Each row returned represents a Terraform version in use. See [View Type: tf_versions](#view-type-tf_versions) for available fields. |
| `providers`   | Information about the Terraform providers in use across the target organization. Each row returned represents a Terraform provider in use. See [View Type: providers](#view-type-providers) for available fields. |
| `modules`     | Information about the Terraform modules in use across the target organization. Each row returned represents a Terraform module in use. See [View Type: modules](#view-type-modules) for available fields. |

> | `resources`   | **Beta.** Information about the Terraform resources tracked across the target organization. Each row returned represents a single resource. See [View Type: resources](#view-type-resources) for available fields. |

Reviewer: ```suggestion
| `resources`   | Information about the Terraform resources tracked across the target organization. Each row returned represents a single resource. This type is in beta. See [View Type: resources](#view-type-resources) for available fields. |
```
We normally reserve boldface for UI labels and to introduce terminology, but I don't think there's a good way to flag beta inline, so feel free to override this suggestion.

| `organization_name`               | `string`   | The name of the organization this workspace belongs to. |
| `project_external_id`             | `string`   | The external ID of the project this workspace belongs to. |
| `project_name`                    | `string`   | The name of the project this workspace belongs to. |
| `provider_count`                  | `number`   | The number of distinct providers used in this workspace. |
| `providers`                       | `string`   | A comma separated list of providers used in this workspace. |
| `resource_count`                  | `number`   | The total number of resources tracked in this workspace's current state. |
| `resources_drifted`               | `number`   | The count of resources that drift was detected for. |
| `resources_undrifted`             | `number`   | The count of resources that drift was not detected for. |
| `source_module_id`                | `string`   | The workspace's source module identifier. |
| `state_version_terraform_version` | `string`   | The Terraform version used to create the current run's state file. |

> | `tags`                            | `string \| null`   | A comma separated list of tags (keys and key:value pairs) applied to this workspace. Returns `null` when no tags are applied. |

Reviewer: ```suggestion
| `tags`                            | `string` or `null`   | A comma separated list of key-value pair tags applied to this workspace. Returns `null` when no tags are applied. |
```

| `source_module_id`                | `string`   | The workspace's source module identifier. |
| `state_version_terraform_version` | `string`   | The Terraform version used to create the current run's state file. |
| `tags`                            | `string \| null`   | A comma separated list of tags (keys and key:value pairs) applied to this workspace. Returns `null` when no tags are applied. |
| `vcs_repo_identifier`             | `string`   | The name of the repository configured for this workspace, if available. |
| `workspace_created_at`            | `datetime` | The time this workspace was created. |
| `workspace_name`                  | `string`   | The name of this workspace. |
| `workspace_terraform_version`     | `string`   | The Terraform version currently configured for this workspace. |
| `workspace_updated_at`            | `datetime` | The time this workspace was last updated. |

#### View type: `tf_versions`

| Field              | Type     | Description |
| ------------------ | -------- | ------------------------------------------------------------------ |

> | `unique_identifier` | `string` | A unique string identifier for this Terraform version entry. |

Reviewer 2: Confusing name, this identifier is not unique and isn't really meant for customer consumption (hence why it's been undocumented up til this point). It's unique across a name/version pair within a state version (or a workspace, if you prefer), but since it's a path like `registry.terraform.io/providers/hashicorp/random/3.7.2`, it can be shared across multiple versions.

This is also something we should figure out how to deprecate, since we'll want to remove it from the API/CSV when we standardize those against the UI later on.

Reviewer 2: Did you have thoughts on how to reword this? I think it's a registry-related path (would need to dig into the details) rather than a true unique identifier.

Author: No, I'm waiting for reply if we should even keep it.

| `source`          | `string` | The source of this module. |
| `version`         | `string` | The semantic version string for this module. |
| `registry_type`   | `string`  | The registry type for this module (for example, `public` or `private`). |
| `curated`         | `boolean` | True if this module is curated, false otherwise. |
| `workspace_count` | `number` | The number of workspaces using this module version. |
| `workspaces`      | `string` | A comma-separated list of workspaces using this module version. |

#### View type: `resources`

-> **Note**: The `resources` view type is in beta and subject to change without notice.

| Field                      | Type       | Description                                                               |
| -------------------------- | ---------- | ------------------------------------------------------------------------- |
| `name`                     | `string`   | The name of this resource. |
| `address`                  | `string`   | The address of this resource. |
| `provider_type`            | `string`   | The resource type identifier as defined by the provider (for example, `aws_instance`, `random_pet`). |

> | `provider_name`            | `string`   | The source address of the provider for this resource. For providers from the public registry, the `registry.terraform.io/` hostname is omitted (for example, `hashicorp/aws`). For built-in providers, the full address is returned (for example, `terraform.io/builtin/terraform`). |

Reviewer 2: This field is just the name of the provider, not the address. We should note that users can get to the address by clicking the name, which will bring them to the Providers view filtered to the appropriate Provider for their resource

| `module_name`              | `string \| null`   | The name of the module for this resource, if available. Returns `null` for top-level resources not inside a module. |
| `terraform_version`        | `string`   | The Terraform version associated with this resource. |
| `billable_rum`             | `boolean`  | True if this resource is a billable resource under management (RUM), false otherwise. |
| `source_type`              | `string`   | The source type for this resource. |
| `source_id`                | `string`   | The source ID for this resource. |

> | `source_updated_at`        | `datetime` | The time this resource's source was last updated. |

Reviewer 2: Can you add some color to what a "source" is? (Happy to help)

Author: If you can find the clean instance in the prod, everything I was looking are fake providers, so example not as good. I've been trying to sync it with API.

| `workspace_name`           | `string`   | The name of the workspace this resource belongs to. |

> | `workspace_external_id`    | `string`   | The external ID of the workspace this resource belongs to. |

Reviewer 2: The external ID fields aren't really meant for user consumption - they're visible in the API/CSV export due to technical limitations, and we have concrete plans to remove after beta. Should they be documented, or should there at least be some signifier that they're deprecated (already)? (Adam and David in particular may have some thoughts on this)

Author: Right now it's exposed and will be returned.
@trujillo-adam Do you have any suggestions for us?

Reviewer: So sorry for the delay responding. Normally, I would say we should document all attributes for completeness, even if the user doesn't do anything with them. Is there a significant gap between GA and the planned removal? It might be less disruptive to remove them now.

| Field              | Type     | Description |
| ------------------ | -------- | ------------------------------------------------------------------ |
| `unique_identifier` | `string` | A unique string identifier for this Terraform version entry. |
| `version`          | `string` | The semantic version string for this Terraform version. |
| `workspace_count`  | `number` | The number of workspaces using this terraform version. |
| `workspaces`       | `string` | A comma-separated list of workspaces using this terraform version. |

#### View Type: `providers`

| Field             | Type     | Description |
| ----------------- | -------- | ----------- |
| `name`            | `string` | The name of this provider. |
| `source`          | `string` | The source of this provider. |
| `version`         | `string` | The semantic version string for this provider. |
| `registry_type`   | `string` | The registry type for this provider (e.g. `public` or `private`). |

> | `curated`         | `boolean`| True if this provider is curated, false otherwise. |

Reviewer: ```suggestion
| `registry_type`   | `string` | The registry type for this provider, such as `public` or `private`. |
| `curated`         | `boolean`| Returns `true` if this provider is curated and `false` otherwise. |
```

| `name`            | `string` | The name of this module. |
| `source`          | `string` | The source of this module. |
| `version`         | `string` | The semantic version string for this module. |
| `registry_type`   | `string`  | The registry type for this module (e.g. `public` or `private`). |

> | `curated`         | `boolean` | True if this module is curated, false otherwise. |

Reviewer: ```suggestion
| `registry_type`   | `string`  | The registry type for this module, such as `public` or `private`. |
| `curated`         | `boolean` | Returns `true` if this module is curated and `false` otherwise. |
```

| `workspace_count` | `number` | The number of workspaces using this module version. |
| `workspaces`      | `string` | A comma-separated list of workspaces using this module version. |

#### View Type: `resources`

-> **Note**: The `resources` view type is currently in beta and subject to change without notice.

| Field                      | Type       | Description                                                               |
| -------------------------- | ---------- | ------------------------------------------------------------------------- |
| `name`                     | `string`   | The name of this resource. |
| `address`                  | `string`   | The address of this resource. |

> | `provider_type`            | `string`   | The resource type identifier as defined by the provider (e.g., `aws_instance`, `random_pet`). |

Reviewer: ```suggestion
| `provider_type`            | `string`   | The resource type identifier as defined by the provider, for example `aws_instance` and `random_pet`. |
```
There is a [style guideline about "e.g."](https://github.com/hashicorp/web-unified-docs/blob/main/docs/style-guide/general/language.md#do-not-use-words-or-phrases-borrowed-from-other-languages-scientific-words-or-jargon-words) and other words. There is one about only using parentheses either to introduce acronyms and as part of a code example, as in a function, but I can't find that one right now.

> | `provider_name`            | `string`   | The source address of the provider for this resource. For providers from the public registry, the `registry.terraform.io/` hostname is omitted (e.g., `hashicorp/aws`). For built-in providers, the full address is returned (e.g., `terraform.io/builtin/terraform`). |

Reviewer: ```suggestion
| `provider_name`            | `string`   | The source address of the provider for this resource. The platform omits the `registry.terraform.io/` hostname when it's from the public registry, for example `hashicorp/aws`. For built-in providers, the platform returns the full address, for example`terraform.io/builtin/terraform`. |
```
Active voice

> | `module_name`              | `string \| null`   | The name of the module for this resource, if available. Returns `null` for top-level resources not inside a module. |

Reviewer: ```suggestion
| `module_name`              | `string` or `null`   | The name of the module for this resource if available. Returns `null` for top-level resources that aren't inside a module. |
```
Just to confirm this is what we mean--that the data type is either a string or `null`, correct?

| `terraform_version`        | `string`   | The Terraform version associated with this resource. |

> | `billable_rum`             | `boolean`  | True if this resource is a billable resource under management (RUM), false otherwise. |

Reviewer: ```suggestion
| `billable_rum`             | `boolean`  | Returns `true` if this resource is a billable resource under management (RUM). The value is `false` otherwise. |
```
To avoid confusion, use backticks to reference actual enums and constructs that appear in the console.

| `lt`               | `number`                      | Filters data to rows where the field value is less than the filter value. |
| `gteq`             | `number`                      | Filters data to rows where the field value is greater than or equal to the filter value. |
| `lteq`             | `number`                      | Filters data to rows where the field value is less than or equal to the filter value. |

> | `is_before`        | `datetime`                    | Filters data to rows where the field value is earlier than the filter value. ISO 8601 formatted timestamps are required for filter values. If no time zone offset is provided, the filter value will be interpreted as UTC. |

Reviewer: ```suggestion
| `is_before`        | `datetime`                    | Filters data to rows where the field value is earlier than the filter value. You must use ISO 8601 formatted timestamps for filter values. If you do not provide a time zone offset, the platform interprets the filter value as UTC. |
```
Active voice

> | `is_after`         | `datetime`                    | Filters data to rows where the field value is earlier than the filter value. ISO 8601 formatted timestamps are required for filter values. If no time zone offset is provided, the filter value will be interpreted as UTC. |

Reviewer: ```suggestion
| `is_after`         | `datetime`                    | Filters data to rows where the field value is earlier than the filter value. You must use ISO 8601 formatted timestamps for filter values. If you do not provide a time zone offset, the platform interprets the filter value as UTC. |
```

  'https://app.terraform.io/api/v2/organizations/$ORG_NAME/explorer?type=modules&filter%5B0%5D%5Bregistry_type%5D%5Bis%5D%5B0%5D=public'
```

**View Type: `resources`**

> _Show all resources_

Reviewer: ```suggestion
The following example call returns all resources:
```
We should update the other example introductions into a complete sentence and regular non-italic font. I'm not sure how they that formatting got into this page.
